[JIT] Make graph->param_node()->next() the first node

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#19788 [JIT] Make graph->param_node()->next() the first node**

We used to use a circular doubly linked list of nodes to implement
Blocks's nodes() list in the JIT. prim::Return held the outputs of
the block, and was also the 'end' node for the list.
prim::Param was not in the list.

This commit tweaks this: The list is still circular and doubly-linked,
but prim::Param is at the beginning of the list, and prim::Return
is at the end. Like before, neither of these nodes show up in the nodes()
iteration. To make sure we never have null edges prim::Return's next is
the prim::Param node, preserving the circular-ness.

Why is this better?

* It is now always safe to ask for prev()/next() on a node.
It will always be non-null.

* It is now possible to always insert a node after
another node, even if the node is a prim::Param. Inserting after prim::Param
puts it at the beginning of the block. This is more consistent with how
things work for the prim::Return, where it is always valid to put something
_before_ it.

Inserting after the return or before the param node is guarded by an assert.

Differential Revision: [D15094260](https://our.internmc.facebook.com/intern/diff/D15094260)

zdevito committed:
[JIT] Make graph->param_node()->next() the first node

We used to use a circular doubly linked list of nodes to implement
Blocks's nodes() list in the JIT. prim::Return held the outputs of
the block, and was also the 'end' node for the list.
prim::Param was not in the list.

This commit tweaks this: The list is still circular and doubly-linked,
but prim::Param is at the beginning of the list, and prim::Return
is at the end. Like before, neither of these nodes show up in the nodes()
iteration. To make sure we never have null edges prim::Return's next is
the prim::Param node, preserving the circular-ness.

Why is this better?

* It is now always safe to ask for prev()/next() on a node.
It will always be non-null.

* It is now possible to always insert a node after
another node, even if the node is a prim::Param. Inserting after prim::Param
puts it at the beginning of the block. This is more consistent with how
things work for the prim::Return, where it is always valid to put something
_before_ it.

Inserting after the return or before the param node is guarded by an assert.

zdevito committed:
Update on "[JIT] Make graph->param_node()->next() the first node"

[JIT] Make graph->param_node()->next() the first node

We used to use a circular doubly linked list of nodes to implement
Blocks's nodes() list in the JIT. prim::Return held the outputs of
the block, and was also the 'end' node for the list.
prim::Param was not in the list.

This commit tweaks this: The list is still circular and doubly-linked,
but prim::Param is at the beginning of the list, and prim::Return
is at the end. Like before, neither of these nodes show up in the nodes()
iteration. To make sure we never have null edges prim::Return's next is
the prim::Param node, preserving the circular-ness.

Why is this better?

* It is now always safe to ask for prev()/next() on a node.
It will always be non-null.

* It is now possible to always insert a node after
another node, even if the node is a prim::Param. Inserting after prim::Param
puts it at the beginning of the block. This is more consistent with how
things work for the prim::Return, where it is always valid to put something
_before_ it.

Inserting after the return or before the param node is guarded by an assert.

gh-metadata: pytorch pytorch 19788 gh/zdevito/32/head

apaszke commented on torch/csrc/jit/graph_node_list.h:
```diff
@@ -114,10 +114,10 @@ struct generic_graph_node_list {
   }
   generic_graph_node_list_iterator<T> end() {
-    return generic_graph_node_list_iterator<T>(head, d);
+    return generic_graph_node_list_iterator<T>(head->next_in_graph[!d], d);
```
This excludes the node prior to `head` from the iteration. It's a bit non-standard for containers to do this, so I'd rather have us slice the `graph_node_list` inside `Graph` than resort to hacks like this.

zdevito replied:
I am not sure exactly what you are asking for here. Can you clarify? 

In this change the node prior to head is always tail. Both head and tail are suppose to be excluded from the list.

Before and after this pass, the node_list always represents the full list and not a slice. It has the property that it cannot be invalidated no matter what inserts/deletes would happen in the list because it doesn't hold a reference to a node in the list.

apaszke reviewed (commented):
This is a nice cleanup, but the semantics of `generic_graph_node_list` got messed up pretty badly. Let's fix it or at least add a comment!!

zdevito committed:
Update on "[JIT] Make graph->param_node()->next() the first node"

[JIT] Make graph->param_node()->next() the first node

We used to use a circular doubly linked list of nodes to implement
Blocks's nodes() list in the JIT. prim::Return held the outputs of
the block, and was also the 'end' node for the list.
prim::Param was not in the list.

This commit tweaks this: The list is still circular and doubly-linked,
but prim::Param is at the beginning of the list, and prim::Return
is at the end. Like before, neither of these nodes show up in the nodes()
iteration. To make sure we never have null edges prim::Return's next is
the prim::Param node, preserving the circular-ness.

Why is this better?

* It is now always safe to ask for prev()/next() on a node.
It will always be non-null.

* It is now possible to always insert a node after
another node, even if the node is a prim::Param. Inserting after prim::Param
puts it at the beginning of the block. This is more consistent with how
things work for the prim::Return, where it is always valid to put something
_before_ it.

Inserting after the return or before the param node is guarded by an assert.

gh-metadata: pytorch pytorch 19788 gh/zdevito/32/head

zdevito committed:
Update on "[JIT] Make graph->param_node()->next() the first node"

[JIT] Make graph->param_node()->next() the first node

We used to use a circular doubly linked list of nodes to implement
Blocks's nodes() list in the JIT. prim::Return held the outputs of
the block, and was also the 'end' node for the list.
prim::Param was not in the list.

This commit tweaks this: The list is still circular and doubly-linked,
but prim::Param is at the beginning of the list, and prim::Return
is at the end. Like before, neither of these nodes show up in the nodes()
iteration. To make sure we never have null edges prim::Return's next is
the prim::Param node, preserving the circular-ness.

Why is this better?

* It is now always safe to ask for prev()/next() on a node.
It will always be non-null.

* It is now possible to always insert a node after
another node, even if the node is a prim::Param. Inserting after prim::Param
puts it at the beginning of the block. This is more consistent with how
things work for the prim::Return, where it is always valid to put something
_before_ it.

Inserting after the return or before the param node is guarded by an assert.

gh-metadata: pytorch pytorch 19788 gh/zdevito/32/head
